### ISSUE TITLE
Allow re-registering of Spark.Variable's

### ISSUE THREAD

**Issue by dmiddlecamp** [spark/firmware#241]
If we allow users to re-register the pointers to their string variables, it will be easier to use, consider this code:

https://github.com/spark/core-firmware/blob/master/src/spark_utilities.cpp#L140-L147

and this example:
```
char foo[24];
int state = 0;

void setup()
{
    Spark.variable("foo", foo, STRING);
}

void loop()
{
    //overwrite or change the pointer for your string

    state = !state;    
    if (state) {
        strcpy(foo, "One");
    }
    else {
        strcpy(foo, "Two");
    }

    counter++;
    delay(5000);
}
```

becomes:
```
char foo[24];
int state = 0;

void setup()
{
    Spark.variable("foo", foo, STRING);
}

void loop()
{
    //overwrite or change the pointer for your string

    state = !state;    
    if (state) {
        foo = "One";
    }
    else {
        foo = "two";
    }
    Spark.variable("foo", foo, STRING);

    counter++;
    delay(5000);
}
```

**Comment by m-mcgowan** [comment]
Fixed for 0.4.0. Variables can be re-registered if they are the same type.

**Comment by tenaciousRas** [comment]
This issue can actually be re-stated as a requirement for Particle.variable(...):  registering a variable causes a pointer to be set, so if the registered variable is set to a new pointer location it must be re-registered.

Maybe update the reference docs at https://docs.particle.io/reference/firmware/core/#particle-variable- about this?  I looked for a way to submit a pull request but I don't think the ref docs are public.
